Connection: Add site creation date argument

tyxla opened the pull request:
As part of our Jetpack Improved Onboarding work, we'd like to send an additional argument with the rest of the connection URL arguments. That argument would contain the site creation date, which is basically the earliest of either: 

* The earliest admin user registration date.
* The earliest post date.

Companion diff: D25012-code.

#### Changes proposed in this Pull Request:

* Introduce a `Jetpack::get_assumed_site_creation_date()` method.
* Add a new `site_created` parameter to the connection URLs.
* Add an exception for multisite subsites, where we use the multisite creation date instead.

#### Testing instructions:

* Follow the testing instructions in D25012-code.
* Checkout this branch.
* Verify tests pass by running either:

```
yarn docker:phpunit --filter test_get_assumed_site_creation_date
```

or:

```
phpunit --filter test_get_assumed_site_creation_date
```

depending on the environment you're developing Jetpack in.

#### Proposed changelog entry for your changes:

* No changelog entry needed.

tyxla committed:
Connection: Add site creation date argument

kraftbj commented:
Would it make sense to query the oldest post object and use the oldest between that and the oldest user?

1. I make a site in 2015 at my company. I was the only person messing with the site.
2. New person is hired in 2018 to help out.
3. I quit in 2019. My user is deleted, authorship changed to user 2.
4. The oldest user is now 2018, even though the site has been around since 2015.

dereksmart commented on class.jetpack.php:
```diff
@@ -4844,6 +4845,24 @@ function build_connect_url( $raw = false, $redirect = false, $from = false, $reg
+	 *
+	 * @return string Site creation date and time.
+	 */
+	public static function get_site_creation_date() {
```
This has enough chance to be a misnomer (related to @kraftbj's comment) that I would recommend we rename it explicitly to `get_earliest_user_registration_date()`

AnnaMag reviewed (commented):
Curious: could we pass a classifier (new vs established site) instead of the actual date— `site_age_category` vs `site_creation_date`? Do we need such detail info to be passed around?

tyxla committed:
Include posts in site creation date assumption

tyxla commented:
Thanks everybody for the feedback!

> Would it make sense to query the oldest post object and use the oldest between that and the oldest user?

Thanks @kraftbj, that scenario makes sense to me. I've addressed that in https://github.com/Automattic/jetpack/pull/11450/commits/0a4a4cbcbc066ac150eef79c39d2732a20f05ec3 - would appreciate feedback on the approach.

> This has enough chance to be a misnomer (related to @kraftbj's comment) that I would recommend we rename it explicitly to get_earliest_user_registration_date()

Thanks @dereksmart - this suggestion made sense, but maybe not after I added the earliest post date in the mix. Just to avoid any confusion you described, I've renamed it to `Jetpack::get_assumed_site_creation_date()`. How does that sound? Do you have any better recommendations?

> Curious: could we pass a classifier (new vs established site) instead of the actual date— site_age_category vs site_creation_date? Do we need such detail info to be passed around?

I think a classifier wouldn't be helpful, because a new site will no longer be a new site after a month, so that classifier could become invalid in some time. Instead, if we store a date, we can do whatever we want with it on the dotcom side (including, automatically maintain an up-to-date classifier as you mentioned with an automated job). I think that the classifier idea is a good idea, but it doesn't make sense to be done here - I'd see it done on the .com side, where we still compute it from the site creation date that we pass here. 

Now that this PR is punted for 7.2, I'd like to add some tests and further manually test it (which is why I marked it as "in-progress". Feedback on the approach is very welcome though!

dereksmart commented:
`get_assumed_site_creation_date` sounds better. 

It may end up growing a bit as we add more "guess factors" to the mix, but if we need to we can split out the method later.

kraftbj commented on class.jetpack.php:
```diff
@@ -7474,3 +7513,5 @@ public static function show_backups_ui() {
 	}
 }
+
+Jetpack::get_assumed_site_creation_date();
```
Is this just a bit of leftover code?

tyxla replied:
Yeah, cleaned up in https://github.com/Automattic/jetpack/pull/11450/commits/8af97d252b0f3dd0db2c60120c0fcbf6af54a82d.

tyxla committed:
Cleanup

tyxla committed:
Be more precise about subsite creation date

tyxla commented on class.jetpack-network.php:
```diff
@@ -476,6 +476,7 @@ public function do_subsiteregister( $site_id = null ) {
 				'state'                 => $user_id,
 				'_ui'                   => $tracks_identity['_ui'],
 				'_ut'                   => $tracks_identity['_ut'],
+				'site_created'          => Jetpack::get_assumed_site_creation_date(),
```
Nice 👍 I like that, implemented in https://github.com/Automattic/jetpack/pull/11450/commits/51989c4f12425f00348c954e848064839be213f2

tyxla committed:
Add some tests for Jetpack::get_assumed_site_creation_date()

tyxla committed:
Remove extra file

tyxla committed:
Update docs

tyxla committed:
Inline docs

tyxla commented:
Thanks everyone for the feedback! This is ready for another review.

AnnaMag reviewed (approved):
LGTM

jeherve reviewed (approved):
This works well for me. 🚢

tyxla commented:
Thank you, folks!
